Store control-system measurement times in the DataBox

wthrowe opened the pull request:
Knowing the times will be necessary for implementing time-step adjustment to avoid deadlocking.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Add control_system::FutureMeasurements

wthrowe committed:
Add FutureMeasurements tag

wthrowe committed:
Initialize FutureMeasurements

wthrowe committed:
Change CHECK to REQUIRE in control Trigger test

CHECK on a non-trivial expression will continue the test if an
assertion fails, which will usually fail several more.  Since often
only the first error is meaningful, terminate on it.

wthrowe committed:
Use FutureMeasurements in control system trigger

knelli2 commented on src/ControlSystem/FutureMeasurements.hpp:
```diff
@@ -0,0 +1,54 @@
+/// \endcond
+
+namespace control_system {
+/// Information about upcoming measurements.
+class FutureMeasurements {
```
Can you expand on this a bit? What could this potentially be used for (what is it intended to be used for)? Maybe say that this is also aware of updates as well as measurements?

wthrowe replied:
I've expanded a little, but I don't really have much to say.  I generally prefer to avoid listing what uses a class in its documentation unless things are really tightly coupled, because that's not a property of the class and can end up out-of-date very easily.

Do you have a suggested wording for a statement about updates?  The "time of an update" seems to be ambiguous, since the SpEC update got split into two things happening at different times and I haven't observed a consensus about which that phrase would refer to.

knelli2 replied:
I think your fixup was a good `\brief` for a class, but it basically just repeated what the namespace and the class name already said: `control_system::FutureMeasurements` = class that knows about future control system measurements. In my opinion that's not super helpful documentation.

I view the docs as me (or an anthropomorphized version of the class) giving a newer dev of the code a (brief) explanation of how this class ties in to the surrounding infrastructure. You don't have to give specific uses of this class like "this class will be used in the control_system::Trigger class for ...". I agree that's hard to maintain. But you can give a broader idea of why this class is useful.

Here's how I would have documented the class. If you decide to use it, feel free to change up some wording. Hopefully the last paragraph removes the ambiguity of what a "time of an update" means.

```
Class for computing the upcoming measurement times for a control
system measurement.

The measurement timescales are stored in the GlobalCache. At a given simulation time,
they can be used to get the current measurement timescale. However, it is
sometimes advantageous to know when future measurements will occur in
case preemptive decisions have to be made. It is difficult to know how to
calculate these future measurements without additional information and
knowledge about the structure of the measurement timescales. This class
has this additional knowledge and takes care of that calculation. See the
`control_system::Tags::MeasurementTimescales` tag for this additional
information.

In addition to knowing about the simulation time of future measurements,
this class also knows the simulation time of control system/FunctionOfTime
updates as well.
```

wthrowe replied:
> I think your fixup was a good `\brief` for a class, but it basically just repeated what the namespace and the class name already said: `control_system::FutureMeasurements` = class that knows about future control system measurements. In my opinion that's not super helpful documentation.

I added that it computes the values, which is an important point that was missing.

> ```
> The measurement timescales are stored in the GlobalCache.
> ```

I don't want to talk about the GlobalCache here.  Avoiding use of the GlobalCache in this class was an intentional decision I made.

> ```
> At a given simulation time,
> they can be used to get the current measurement timescale. However, it is
> sometimes advantageous to know when future measurements will occur in
> case preemptive decisions have to be made.
> ```

These two sentences are good.  I'll use something like them.

> ```
> It is difficult to know how to
> calculate these future measurements without additional information and
> knowledge about the structure of the measurement timescales. This class
> has this additional knowledge and takes care of that calculation. See the
> `control_system::Tags::MeasurementTimescales` tag for this additional
> information.
> ```

It's not factored out because it's difficult or confusing.  It's just `t + timescale(t)`.  It's factored out because it seems like a bad idea to calculate and store all the values in two places.

Although, now that I poke around, I don't actually see `t + timescale(t)` documented anywhere, and this does seem like a pretty good place for it.

Linking to the tag is a good idea, too.

>  ``` 
> In addition to knowing about the simulation time of future measurements,
> this class also knows the simulation time of control system/FunctionOfTime
> updates as well.
> ```

That's actually the "time of update" that it doesn't know about.  It knows about the measurement half.

knelli2 commented on src/ControlSystem/FutureMeasurements.hpp:
```diff
@@ -0,0 +1,54 @@
+  FutureMeasurements(size_t measurements_per_update,
+                     double first_measurement_time);
+
+  // Next measurement time, if known.
+  std::optional<double> next_measurement() const;
+  // Next measurement that triggers an update, if known.
```
Doxygen for both of these?

knelli2 commented on src/ControlSystem/FutureMeasurements.cpp:
```diff
@@ -0,0 +1,65 @@
+FutureMeasurements::FutureMeasurements(const size_t measurements_per_update,
+                                       const double first_measurement_time)
+    : measurements_{std::numeric_limits<double>::signaling_NaN(),
+                    first_measurement_time},
+      measurements_until_update_(measurements_per_update - 1),
```
```cpp
ASSERT(measurements_per_update > 0, ...)
```
? Obviously other things should break if this was actually false, but would be good as a safety check.

knelli2 commented on tests/Unit/ControlSystem/Actions/Test_InitializeMeasurements.cpp:
```diff
@@ -234,11 +257,25 @@ SPECTRE_TEST_CASE("Unit.ControlSystem.InitializeMeasurements",
+}
+
+SPECTRE_TEST_CASE("Unit.ControlSystem.InitializeMeasurements",
+                  "[ControlSystem][Unit]") {
+  test_initialize_measurements(true, true);
```
[optional] I know it's the same number of lines, but it's more easy to generalize in the future and it's clear what the two arguments are for.
```cpp
  for (const auto& [ab_active, c_active] :
       cartesian_product(make_array(true, false), make_array(true, false))) {
    test_initialize_measurements(ab_active, c_active);
  }
```

wthrowe replied:
I prefer the current version.

(And why `make_array(true, false)` instead of `std::array{true, false}`?)

knelli2 replied:
No real reason. Just copied this from elsewhere that used the `cartesian_product`

knelli2 commented on src/ControlSystem/Trigger.hpp:
```diff
@@ -83,125 +75,92 @@ class Trigger : public DenseTrigger {
-    const bool triggered = time == *next_trigger_;
-    if (triggered) {
-      *next_trigger_ = next_measurement(time, measurement_timescales);
+    if (not measurement_times->next_measurement().has_value()) {
```
Will this work? Currently, every element has to wait for every FoT to be ready. Here, you're only checking the functions of time when these specific control systems need to be updated. But what about the other control systems? If we only check these specific FoTs on this element, then I believe your current implementation would work.

wthrowe replied:
These are measurement timescales, not FoTs.

knelli2 reviewed (commented):
You can squash the docs changes directly if you decide to make them.

wthrowe commented:
Pushed fixups.  Changes:
* Documentation update discussed above.  I decided to put the update formula on the `update` method.
* Changed inactive control systems to report one measurement per update.  I found this more convenient for code in a later PR.
* Changed the assertion about the list growing long to a non-error limit on the list size.  The class never uses more than the first cycle, but storing a bit more means no thought has to go into the exact number of entries required.  I also got worried that it might try to store every measurement time during a wave-zone step in LTS.

knelli2 reviewed (commented):
Looks good. Squash

knelli2 commented:
Not sure what some of the test failures are because I can't see the output. But they failed building the executables

wthrowe commented:
For gcc-9 Release, "View raw logs" shows "g++-9: fatal error: Killed signal terminated program cc1plus", so presumably OOM.  Doesn't work for gcc-10 Release, though.

knelli2 commented:
Ok sounds good. Nothing to block this PR
